routing/index_graph.cpp

```diff
@@ -1,10 +1,115 @@
-#include "index_graph.hpp"
+#include "routing/index_graph.hpp"
+
+#include "routing/restrictions_serialization.hpp"
 
 #include "base/assert.hpp"
 #include "base/exception.hpp"
 
+#include "std/algorithm.hpp"
 #include "std/limits.hpp"
 
+namespace
+{
+using namespace routing;
+
+bool IsRestricted(RestrictionVec const & restrictions, vector<SegmentEdge> const & edges,
+                  Segment const & u, Segment const & v, bool isOutgoing)
+{
+  uint32_t const featureIdFrom = isOutgoing ? u.GetFeatureId() : v.GetFeatureId();
+  uint32_t const featureIdTo = isOutgoing ? v.GetFeatureId() : u.GetFeatureId();
+
+  // Looking for at least one restriction of type No from |featureIdFrom| to |featureIdTo|.
+  if (binary_search(restrictions.cbegin(), restrictions.cend(),
+                    Restriction(Restriction::Type::No, {featureIdFrom, featureIdTo})))
+  {
+    if (featureIdFrom != featureIdTo)
+      return true;
+
+    // @TODO(bykoianko) According to current code if a feature id is marked as a feature with
+    // restrictricted uturn it's restricted to make a uturn on the both ends of the feature.
+    // Generally speaking it's wrong. In osm there's information about the end of the feature
+    // where the uturn is restricted. It's necessary to pass the data to mwm and to use it here.
+    // Set test LineGraph_RestrictionF1F1No for further details.
+    return u.GetSegmentIdx() == v.GetSegmentIdx() && u.IsForward() != v.IsForward();
```

Comment:
To answer this comment I need make several tests:
- to check how it's efficient to get rid of U-turns on intermediate feature points
- to check if it's worth exchanging restriction with type No with restrictions with type Only
So, I'll answer later probably the code will be rewritten.